crates/api/src/config.rs

```diff
@@ -14,68 +15,236 @@ fn tc<S: serde::Serialize, D: serde::de::DeserializeOwned>(
+                                return Err(K2Error::other(format!(
+                                "{debug_path} attempted to insert a map where an entry exists",
+                            )));
+                            }
+                        }
+                    }
+                    ConfigMap::ConfigEntry(new_entry) => match old_map {
+                        ConfigMap::ConfigMap(m) => {
+                            if !m.is_empty() {
+                                return Err(K2Error::other(format!(
+                                    "{debug_path} attempted to insert an entry where a map exists",
+                                )));
+                            }
+                            *old_map =
+                                ConfigMap::ConfigEntry(new_entry.clone());
+                            if is_runtime {
+                                tracing::warn!(debug_path, "no update callback for runtime config alteration");
+                            }
+                        }
+                        ConfigMap::ConfigEntry(old_entry) => {
+                            old_entry.value = new_entry.value.clone();
+                            if let Some(update_cb) = &old_entry.update_cb {
+                                updates.push((
+                                    update_cb.clone(),
+                                    new_entry.value.clone(),
+                                ));
+                            } else if is_runtime {
+                                tracing::warn!(debug_path, "no update callback for runtime config alteration");
+                            }
+                        }
+                    },
+                }
+                Ok(())
+            }
+            apply_map(
+                &debug_path,
+                are_defaults_set,
+                is_runtime,
+                &mut updates,
+                old_map,
+                new_map,
+            )?;
+        }
+        for (update_cb, value) in updates {
+            update_cb(value);
         }
-        self.0.insert(module_name, tc(&M::default())?);
         Ok(())
     }
 
-    /// When kitsune2 is initializing, it will call the factory function
-    /// for all of its modules with an immutable reference to this config
-    /// struct. Each of those modules may choose to call this function
-    /// to extract a module config. Note that this config is loaded from
-    /// disk and can be edited by humans, so the serialization on the module
-    /// config should be tolerant to missing properties, setting sane defaults.
-    /// A module may choose to warn about missing properties and should warn about extraneous properties.
-    pub fn get_module_config<M: ModConfig>(
+    /// Call this in your module constructor once for every parameter for
+    /// which you would like to receive runtime updates. This will immediately
+    /// invoke the callback with the current value to ensure this is atomic.
```

Comment:
```suggestion
    /// invoke the callback with the current value to ensure this is idempotent.
```

?

Reply:
No, I meant atomic. Because another thread could be setting the value at the time you're registering the update cb. If it immediately calls it with the current value when registering the update callback you have the option to *only* pay attention to values sent through the callback.

Reply:
I see! Right yeah I hadn't thought about it like that. Cool, yeah this makes sense then